Restore thread_local states in continuation thread on RPC servers

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38512 Restore thread_local states in continuation thread on RPC servers**

As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

Commit:
Restore thread_local states in continuation thread on RPC servers

As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

[ghstack-poisoned]

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2101,6 +2101,43 @@ def post_hook_add_two(output_grads, input_grads):
+        return t3
+
+    @dist_init
+    def test_thread_local_context_id(self):
```
I confirm that this test fails without this fix.

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -469,7 +473,22 @@ void RequestCallbackImpl::processRpc(
-           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture)]() {
+           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture),
+           threadLocalState = ThreadLocalState(),
+           ctxId = autogradContext->contextId()]() {
```
This is currently the only place on server-side where we have such a continuation by continuing processing with `addCallback`, correct?

mrshenli replied:
There are more in other types of requests. But IIUC, this is the only place we need to propagate context id. 

* In BACKWARD_AUTOGRAD_REQ, it does not need the context id when creating the `PropagateGradientsResp` obj
* In SCRIPT_CALL and SCRIPT_REMOTE_CALL, the context id is fixed by #36395.
* PYTHON_CALL and PYTHON_REMOTE_CALL currently do not yield. But when we add async user function, we also need to propagate the TLS here.
* SCRIPT_RREF_FETCH_CALL and PYTHON_RREF_FETCH_CALL do not need the context id either.

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -469,7 +473,22 @@ void RequestCallbackImpl::processRpc(
            fromWorkerId,
-           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture)]() {
+           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture),
+           threadLocalState = ThreadLocalState(),
```
Do we need TLS state or just the dist autograd context id for now? (I am planning to eventually use TLS state for distributed profiler work, but curious if this is already needed now)

rohan-varma replied:
Also, could we potentially reuse the approach taken in `record_function_ops.cpp`?
It basically does:
```
// Save and pass thread local state into the callback
  at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
// restore tls_state in lambda
at::ThreadLocalStateGuard g(tls_state);
});
```

The reasoning for this is that using `ThreadLocalStateGuard` is preferred to directly accessing `ThreadLocalState` as I understand it, after some discussion with @ilia-cher

mrshenli replied:
> Do we need TLS state or just the dist autograd context id for now? (I am planning to eventually use TLS state for distributed profiler work, but curious if this is already needed now)

For now we only need the context id I think. @xush6528 also pointed out that we will need this for profiler later, so added `ThreadLocalState` as well. I think we can also remove it in this PR and leave to the profiler-related PR?

> This is basically the same here, but there we declare the tls_state outside of the lambda capture and std::move it into the capture. Not sure if there's a difference perf wise.

Should be the same I guess, as both are rvalue references?

xush6528 replied:
You can remove. I think profiler state is not needed because no more tensor operations happens in this callback.
Autograd context id is needed because getMessageWithAutograd(..) needs it.

We will need to restore both profiler state and autograd context id for python call user async function continuation.

xush6528 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2101,6 +2101,43 @@ def post_hook_add_two(output_grads, input_grads):
+
+        with dist_autograd.context() as context_id:
+            loss = rref.to_here().sum()
+            # due to slow add, the continuatoin of this backward pass will be
```
continuatoin ->

mrshenli replied:
Thanks! Will fix when other comments (if there are any). Trying to avoid kicking of CI again for a comment fix.

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

[ghstack-poisoned]

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

[ghstack-poisoned]

xush6528 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2101,6 +2101,43 @@ def post_hook_add_two(output_grads, input_grads):
+        rref = rpc.remote(dst, DistAutogradTest._slow_add, args=(t1, t2))
+
+        with dist_autograd.context() as context_id:
+            loss = rref.to_here().sum()
```
Let me write down some notes for someone who is like me not super clear with FORWARD_AUTOGRAD_REQ.

The `rref.to_here()` here is a `PYTHON_RREF_FETCH_CALL`, which is always wrapped in `FORWARD_AUTORAD_REQUEST`, because of `forceGradRecording` on this line.

https://github.com/pytorch/pytorch/blob/dfcea820df987cb12a68659c15747ee32e31ddad/torch/csrc/distributed/rpc/rref_impl.cpp#L145-L149

When the `PYTHON_RREF_FETCH_RESPONSE` is ready, as in this line.

https://github.com/pytorch/pytorch/blob/dfcea820df987cb12a68659c15747ee32e31ddad/torch/csrc/distributed/rpc/request_callback_impl.cpp#L401

The response message be wrapped into `FORWARD_AUTOGRAD_RESP` by this line. 

https://github.com/pytorch/pytorch/blob/dfcea820df987cb12a68659c15747ee32e31ddad/torch/csrc/distributed/rpc/request_callback_impl.cpp#L496-L499

Where the critical `getMessageWithAutograd(...)` is called to add a SendFunction to the autograd context and the on the client receiving this `FORWARD_AUTOGRAD_RESP ` will add a corresponding RecvFunction to the dist autograd context.

If the autograd context is not restored, the SendFunction will be added to a wrong autograd context. and the following `sum()` backward will be propagated back to server but in a wrong autograd context. 

Why this test can reproduce the thread switch?

By making `rpc.remote(..., _slow_add, ...)` request slow, and the `rref.to_here()` here is processed before the RRef value is set.

So the `to_here()` request callback will run on another thread. The thread to run it is exactly the thread for `rpc.remote(..)`, because it marks the value as ready, thus responsible to run the added callbacks. 
Since `rpc.remote(..)` is called outside of a dist autograd context, `getMessageWithAutograd(...)` does not wrap the `PYTHON_REMOTE_CALL` into a `FORWARD_AUTORAD_REQUEST`, no autograd context is available on this thread.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2101,6 +2101,43 @@ def post_hook_add_two(output_grads, input_grads):
+            # thread_local states properly when jumping across threads on the
+            # server side.
+            dist_autograd.backward(context_id, [loss])
+            grads = dist_autograd.get_gradients(context_id)
```
ah, yes, thanks

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

[ghstack-poisoned]

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

[ghstack-poisoned]

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -469,7 +472,20 @@ void RequestCallbackImpl::processRpc(
+            // thread_local states there.
+            // TODO: Land on a general solution for RPC ThreadLocalState. See
+            // https://github.com/pytorch/pytorch/issues/38510
+            DistAutogradContextGuard ctxGuard(ctxId);
```
Shouldn't we do this in `addCallback` itself instead of fixing each callsite of addCallback?

mrshenli replied:
Yes, eventually, we should do that I think. But for now we still have two Futures (utils and ivalue), and not every callback needs to capture ThreadLocalState, it might be better to first fix master for now. Let's revisit this when we reach a consensus on how we should implement RPC ThreadLocalState.

Commit:
Update on "Restore thread_local states in continuation thread on RPC servers"


As we gradually making the RPC non-blocking on server side, the
processing of the same request can yield-run on different threads.
Hence, we need to populate thread_local states (e.g., ctx id) in
the continuation thread.

Fixes #38439

Differential Revision: [D21583642](https://our.internmc.facebook.com/intern/diff/D21583642)

[ghstack-poisoned]

mrshenli commented:
If we are going to release v1.5.1, we should include this fix.
